### PR TITLE
Support Kotlin suspend functions as scheduler methods

### LINKED ISSUE #21428: Support kotlin coroutines in quarkus-scheduler
### Description
`quarkus-scheduler` does not support suspend functions. 

In the past i bridged the gap with custom code, but i think it should be supported by quarkus.

I had this requirement multiple times. 

```kotlin
@ApplicationScoped
class SomeBean {

    @Scheduled(every="10s")     
    suspend fun doSomething() {
        // ...
    }
}
```

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by geoand**
I had to juggle around some of the classes and introduce new modules in order to introduce Kotlin support in the same way we have done it for reactive messaging and in order to avoid split package issues.

Resolves: #21428

### CODE REVIEW

**Comment by geoand** [comment]
This is still in draft because I need to add some test and potentially clean up the code a bit.

**Comment by mkouba** [comment]
> I had to juggle around some of the classes and introduce new modules in order to introduce Kotlin support in the same way we have done it for reactive messaging and in order to avoid split package issues.

Hm, can't `quarkus-scheduler-kotlin` just depend on `quarkus-scheduler`? It's not even an extension, or? I don't understand why the `common` and `api` modules are needed?

**Comment by geoand** [comment]
> Hm, can't quarkus-scheduler-kotlin just depend on quarkus-scheduler

It can, but that means that users will have to pull it in explicitly. 
The way it's done now, all users have to do is include `quarkus-kotlin` and `quarkus-scheduler`

**Comment by mkouba** [comment]
> > Hm, can't quarkus-scheduler-kotlin just depend on quarkus-scheduler
> 
> It can, but that means that users will have to pull it in explicitly. The way it's done now, all users have to do is include `quarkus-kotlin` and `quarkus-scheduler`

I see. What magic does include the relevant extensions? ;-)

**Comment by geoand** [comment]
In this case there is no magic, it's just that `quarkus-scheduler` brings in `quarkus-scheduler-kotlin`, but that one has optional dependencies on Kotlin, which are basically met when `quarkus-kotlin` is included.

We have used the same exact strategy in Kotlin support for Reactive Messaging. In that case we didn't need to introduce these `api` and `common` modules split, because the API is in RM and could be references by without changes.

**Comment by mkouba** [comment]
> In this case there is no magic, it's just that `quarkus-scheduler` brings in `quarkus-scheduler-kotlin`, but that one has optional dependencies on Kotlin, which are basically met when `quarkus-kotlin` is included.

Couldn't we just add the optional dependencies on Kotlin to `quarkus-scheduler`? (I'm sorry if it's a stupid question ;-)

**Comment by geoand** [comment]
> Couldn't we just add the optional dependencies on Kotlin to quarkus-scheduler? (I'm sorry if it's a stupid question ;-)

Not stupid at all!

We certainly could, but then that would end up creating problems for anyone wanting to work on `quarkus-scheduler` in an IDE that doesn't really support Kotlin (we had this problem in RESTEasy Reactive where Stef was using Eclipse and got to a point where he couldn't work on it until we moved the Kotlin stuff to a separate module)

**Comment by mkouba** [comment]
> > Couldn't we just add the optional dependencies on Kotlin to quarkus-scheduler? (I'm sorry if it's a stupid question ;-)
> 
> Not stupid at all!
> 
> We certainly could, but then that would end up creating problems for anyone wanting to work on `quarkus-scheduler` in an IDE that doesn't really support Kotlin (we had this problem in RESTEasy Reactive where Stef was using Eclipse and got to a point where he couldn't work on it until we moved the Kotlin stuff to a separate module)

Makes sense.

**Comment by geoand** [comment]
PR updated to fix a conflict

**Comment by geoand** [comment]
> PR updated to fix a conflict

And now another one

File: extensions/scheduler/runtime/src/test/java/io/quarkus/scheduler/runtime/util/SchedulerUtilsTest.java [hunk 1]
```diff
@@ -3,6 +3,8 @@
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
+import io.quarkus.scheduler.common.runtime.util.SchedulerUtils;
```
Comment (mkouba) [inline extensions/scheduler/runtime/src/test/java/io/quarkus/scheduler/runtime/util/SchedulerUtilsTest.java:6]: This test should go in the `common` module.

**Review by mkouba — COMMENTED** [review]
Looks good but it would be great if some kotlin expert could review the kotlin parts. CC @evanchooly

File: extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/SchedulerProcessor.java [hunk 2]
```diff
@@ -139,9 +139,11 @@ void collectScheduledMethods(BeanArchiveIndexBuildItem beanArchives, BeanDiscove
             }
             MethodInfo method = annotationInstance.target().asMethod();
             if (Modifier.isStatic(method.flags())) {
-                scheduledBusinessMethods.produce(new ScheduledBusinessMethodItem(null, method, schedules,
-                        transformedAnnotations.getAnnotation(method, SchedulerDotNames.NON_BLOCKING) != null));
-                LOGGER.debugf("Found scheduled static method %s declared on %s", method, method.declaringClass().name());
+                if (!KotlinUtil.isSuspendMethod(method)) {
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/SchedulerProcessor.java]: Done

File: extensions/scheduler/runtime/src/test/java/io/quarkus/scheduler/runtime/util/SchedulerUtilsTest.java (hunk 1 above)
Comment (geoand) [reply extensions/scheduler/runtime/src/test/java/io/quarkus/scheduler/runtime/util/SchedulerUtilsTest.java:6]: Good catch, fixed

File: extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/SchedulerProcessor.java [hunk 3]
```diff
@@ -553,4 +601,17 @@ UnremovableBeanBuildItem unremoveableSkipPredicates() {
         return new UnremovableBeanBuildItem(new UnremovableBeanBuildItem.BeanTypeExclusion(SchedulerDotNames.SKIP_PREDICATE));
     }
 
+    @BuildStep
+    void produceCoroutineScope(BuildProducer<AdditionalBeanBuildItem> buildItemBuildProducer) {
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/SchedulerProcessor.java:603]: Yeah, we should probably do that across Quarkus. But probably something for a follow up PR

**Review by mkouba — APPROVED** [review]
Looks good.

**Pull request merged**